### ISSUE TITLE
Non-PS Eye cameras.

### ISSUE THREAD

**Issue by craftxbox** [cboulay/PSMoveService#534]
Do they work at all at this moment? is there a way to get psmoveservice to use them right now?

**Comment by cboulay** [comment]
Non PS Eye cameras are hard-coded to be disabled because users were complaining that their laptop webcams or whatever cameras were being engaged when they didn't want them to be. We never implemented a camera selection toggle in the GUI so instead we just disabled non PS Eye cameras.

For your own sake, you can try uncommenting [this code](https://github.com/cboulay/PSMoveService/blob/master/src/psmoveservice/PSMoveTracker/PSEye/PSEyeVideoCapture.cpp#L525-L529) and see if it works.

If you're so inclined then we'd also appreciate a pull request that allows toggling specific cameras. This is actually a lot more work than you might think.

Sorry for the late reply. If this doesn't help you then hopefully someone else will get use of it.

**Comment by Wishdream** [comment]
Let's see if I could take a shot at this.
Especially since I know OpenCV and SDL2. Should be a good challenge.

Is the GUI part inside psmoveconfigtool and at what part should a camera selection toggled be implemented in the GUI?

**Comment by HipsterSloth** [comment]
@Wishdream 

> Is the GUI part inside psmoveconfigtool? 

Yes the configuration GUI is all inside of the PSMoveConfigTool. The "tracker settings" portion specifically.

> At what part should a camera selection toggled be implemented in the GUI?

This is going to be a long answer so sorry for the wall of text.

OpenCV also has it's own wrapper for generic cameras but there are a few shortcomings that make this problematic:

* There is no way to get a camera identifier. 
  * You can only open a camera by index. 
  * This makes it a bit of a hassle to store camera calibration values reliably.
* Not all of the WMF camera parameters are exposed through the OpenCV interface
* There is no good way to query camera capabilities through the OpenCV interface
  * What resolutions does the camera support?
  * What frame rates does the camera support?
  * What buffer formats the camera support?
  ...

If you were to the OpenCV camera interface for this you would need the user to specify the number of cameras and what there settings are. This could either be done by making a set of JSON config files or editing them UI and saving them out. However I wouldn't recommend this approach.

I should mention some work that has since been done in my [PSVR testbed project](https://github.com/HipsterSloth/PSVRTracker) related to non-PSEye cameras. In that project I've implemented support for generic webcams through the Window Media Foundation and a generalize system for tracker configuration. See [WMFTracker folder](https://github.com/HipsterSloth/PSVRTracker/tree/master/src/psvrservice/WMFTracker) in the PSVRTracker project. 

The other big issue you have to contend with once you have a video stream opened is lens calibration. You need to know what the lens distortion and focal lengths are for a given camera so that you can accurately estimate bulb position relative to the camera. This is done with the [OpenCV camera calibration API](https://docs.opencv.org/3.1.0/dc/dbb/tutorial_py_calibration.html) and the checkerboard pattern. If you have played with OpenCV before chances are you have experienced how difficult it can be to get a reliable lens calibration. The previous version of the lens calibration tool I had in the PSMoveConfigTool would often give unreliable results, which is mainly why we didn't support anything but the PS3Eye camera since trying to help people debug their lens calibrations would be a nightmare.

I finally got the lens calibration working pretty reliably in the PSVR tracker testbed. See [AppStage_MonoCalibration.cpp](https://github.com/HipsterSloth/PSVRTracker/blob/master/src/psvrconfigtool/AppStage_MonoCalibration.cpp). The main key to reliable calibration seemed to be holding fixed several of the distortion parameters that aren't really needed for normal web cameras. See [line 654](https://github.com/HipsterSloth/PSVRTracker/blob/master/src/psvrconfigtool/AppStage_MonoCalibration.cpp#L654-L666).

It was my intention to port over all of this camera work into PSMoveService after I had finished fixing a few bugs with camera pose estimation in PSVRTracker, but I've been swamped with trying to get the latest release of PSMoveSteamVRBridge into a better place.

If this still sounds like something you would still like to help port over I would love the help and am happy to answer any questions about code in either PSMoveService or in PSVRTracker.

**Comment by Wishdream** [comment]
@HipsterSloth 
I actually appreciate the wall of text in this case. Since it does give me quite an overview of how much has been done and how much needs to be made.

> OpenCV also has it's own wrapper for generic cameras but there are a few shortcomings that make this problematic:

I used to have an external library for getting camera parameters for each camera and use them to store information and compare them to OpenCV. It's not a sound solution though since it's possible to have two cameras that have the same parameters. I'll have to play around with the tools and open source libraries I have to see what works best. I wish it was as easy on Linux though since they have V4L which makes everything infinitely easier.

I'll look into what you have going with WMF though. May help me in producing a better system for generic cameras. And hopefully one that is cross platform.

> The other big issue you have to contend with once you have a video stream opened is lens calibration.

I also have messed with OpenCV's camera calibration API. I have dealt with this with my traffic thesis before and you are right but I have also made it work reliably as well. I'll have to check your code and compare with what I have currently just to make sure.

I may have to do it in PSVRTracker first as well before porting it. I think it should be much easier to work with less overhead until it's ready for PSMoveService.

I'd be happy to help out with this, and I'll let you know if I need some help.
